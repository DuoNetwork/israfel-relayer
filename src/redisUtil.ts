// import Redis from 'ioredis';
// import * as CST from '../common/constants';
// import { IOrderBook, ITrade } from '../common/types';
// import util from './util';

// export class RedisUtil {
// 	private redisPub: Redis.Redis | null = null;
// 	private redisSub: Redis.Redis | null = null;
// 	private handleOrderBooksUpdate:
// 		| ((channel: string, orderBooks: { [src: string]: IOrderBook }) => any)
// 		| null = null;
// 	private handleTradesUpdate: ((channel: string, trades: ITrade[]) => any) | null = null;

// 	public init(redisKey: { host: string; password: string; servername: string }) {
// 		this.redisPub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
// 		this.redisSub.on('pmessage', (pattern, channel, message) =>
// 			this.onMessage(channel, message, pattern)
// 		);
// 	}

// 	public onMessage(channel: string, message: string, pattern: string = '') {
// 		util.logDebug(pattern + channel);
// 		const [type, ...others] = channel.split('|');
// 		switch (type) {
// 			case CST.DB_ORDER_BOOKS:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, { [others[0]]: JSON.parse(message) });
// 				break;
// 			case CST.DB_OB_OVERVIEW:
// 			case CST.DB_OB_GLOBAL:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, JSON.parse(message));
// 				break;
// 			case CST.DB_TRADES:
// 				if (this.handleTradesUpdate) this.handleTradesUpdate(channel, JSON.parse(message));
// 				break;
// 			default:
// 				break;
// 		}
// 	}

// 	public onOrderBooks(
// 		handleOrderBooksUpdate: (channel: string, orderBooks: { [src: string]: IOrderBook }) => any
// 	) {
// 		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
// 	}

// 	public onTrades(handleTradesUpdate: (channel: string, trades: ITrade[]) => any) {
// 		this.handleTradesUpdate = handleTradesUpdate;
// 	}

// 	public publish(channel: string, msg: string): void {
// 		if (this.redisPub) this.redisPub.publish(channel, msg);
// 	}

// 	public subscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.subscribe(channel);
// 	}

// 	public patternSubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.psubscribe(pattern);
// 	}

// 	public unsubscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.unsubscribe(channel);
// 	}

// 	public patternUnsubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.punsubscribe(pattern);
// 	}

// 	public push(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.lpush(key, values);
// 	}

// 	public putBack(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.rpush(key, values);
// 	}

// 	public async pop(key: string): Promise<string> {
// 		if (this.redisSub) return await this.redisSub.rpop(key);
// 		return '';
// 	}

// 	public async getQueueLength(key: string): Promise<number> {
// 		if (this.redisSub) return this.redisSub.llen(key);
// 		return 0;
// 	}
// }

// const redisUtil = new RedisUtil();
// export default redisUtil;
