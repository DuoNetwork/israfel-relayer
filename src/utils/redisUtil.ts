import Redis from 'ioredis';
import * as CST from '../common/constants';
import {  IOrderBookSnapshotUpdate, IOrderQueueItem } from '../common/types';
import util from './util';

class RedisUtil {
	private redisPub: Redis.Redis | null = null;
	private redisSub: Redis.Redis | null = null;
	private handleOrderBookUpdate:
		| ((channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any)
		| null = null;
	private handleOrderUpdate:
		| ((channel: string, orderQueueItem: IOrderQueueItem) => any)
		| null = null;

	public init(redisKey: { host: string; password: string; servername: string }) {
		this.redisPub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
		this.redisSub.on('pmessage', (pattern, channel, message) =>
			this.onMessage(channel, message, pattern)
		);
	}

	public onMessage(channel: string, message: string, pattern: string = '') {
		util.logDebug(pattern + channel + message);
		const type = channel.split('|')[0];
		switch (type) {
			case CST.DB_ORDER_BOOKS:
				if (this.handleOrderBookUpdate)
					this.handleOrderBookUpdate(channel, JSON.parse(message));
				break;
			case CST.DB_ORDERS:
				if (this.handleOrderUpdate) this.handleOrderUpdate(channel, JSON.parse(message));
				break;
			default:
				break;
		}
	}

	public onOrderBookUpdate(
		handleOrderBookUpdate: (channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any
	) {
		this.handleOrderBookUpdate = handleOrderBookUpdate;
	}
	public onOrderUpdate(
		handleOrderUpdate: (channel: string, orderQueueItem: IOrderQueueItem) => any
	) {
		this.handleOrderUpdate = handleOrderUpdate;
	}

	public publish(channel: string, msg: string) {
		if (this.redisPub) return this.redisPub.publish(channel, msg);
		return Promise.resolve(0);
	}

	public increment(key: string) {
		if (this.redisPub) return this.redisPub.incr(key);
		return Promise.reject();
	}

	public set(key: string, value: string) {
		if (this.redisPub) return this.redisPub.set(key, value);
		return Promise.resolve('');
	}

	public get(key: string) {
		if (this.redisPub) return this.redisPub.get(key);
		return Promise.resolve('');
	}

	public hashSet(key: string, field: string, value: string) {
		if (this.redisPub) return this.redisPub.hset(key, field, value);
		return Promise.resolve(0);
	}

	public hashGet(key: string, field: string): Promise<string | null> {
		if (this.redisPub) return this.redisPub.hget(key, field);
		return Promise.resolve(null);
	}

	public async hashMultiGet(key: string, ...fields: string[]) {
		if (this.redisPub) {
			const values: Array<string | null> = await this.redisPub.hmget(key, ...fields);
			const output: { [field: string]: string | null } = {};
			fields.forEach((f, i) => (output[f] = values[i]));
			return output;
		}
		return Promise.resolve({});
	}

	public hashGetAll(key: string): Promise<any> {
		if (this.redisPub) return this.redisPub.hgetall(key);
		return Promise.resolve(null);
	}

	public hashDelete(key: string, field: string) {
		if (this.redisPub) return this.redisPub.hdel(key, field);
	}

	public subscribe(channel: string) {
		if (this.redisSub) this.redisSub.subscribe(channel);
	}

	public patternSubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.psubscribe(pattern);
	}

	public unsubscribe(channel: string) {
		if (this.redisSub) this.redisSub.unsubscribe(channel);
	}

	public patternUnsubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.punsubscribe(pattern);
	}

	public push(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.lpush(key, values);
	}

	public putBack(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.rpush(key, values);
	}

	public pop(key: string) {
		if (this.redisPub) return this.redisPub.rpop(key);
		return Promise.resolve('');
	}

	public multi() {
		if (this.redisPub) return this.redisPub.multi({ pipeline: false });
		return Promise.resolve('');
	}

	public exec() {
		if (this.redisPub) return this.redisPub.exec();
		return Promise.resolve('');
	}

	public getQueueLength(key: string) {
		if (this.redisPub) return this.redisPub.llen(key);
		return Promise.resolve(0);
	}
}

const redisUtil = new RedisUtil();
export default redisUtil;
