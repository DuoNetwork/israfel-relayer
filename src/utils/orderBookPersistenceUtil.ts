import * as CST from '../common/constants';
import {
	IOrderBookSnapshot,
	IOrderBookSnapshotUpdate
} from '../common/types';
import redisUtil from './redisUtil';
import util from './util';

class OrderBookPersistenceUtil {
	public subscribeOrderBookUpdate(
		pair: string,
		handleOrderBookUpdate: (
			channel: string,
			orderBookSnapshotUpdate: IOrderBookSnapshotUpdate
		) => any
	) {
		redisUtil.onOrderBookUpdate(handleOrderBookUpdate);
		redisUtil.subscribe(`${CST.DB_ORDER_BOOKS}|${CST.DB_UPDATE}|${pair}`);
	}

	public unsubscribeOrderBookUpdate(pair: string) {
		redisUtil.unsubscribe(`${CST.DB_ORDER_BOOKS}|${CST.DB_UPDATE}|${pair}`);
	}

	public async publishOrderBookUpdate(
		pair: string,
		orderBookSnapshot: IOrderBookSnapshot,
		orderBookSnapshotUpdate?: IOrderBookSnapshotUpdate
	): Promise<boolean> {
		try {
			await redisUtil.set(
				`${CST.DB_ORDER_BOOKS}|${CST.DB_SNAPSHOT}|${pair}`,
				JSON.stringify(orderBookSnapshot)
			);
			if (orderBookSnapshotUpdate)
				await redisUtil.publish(
					`${CST.DB_ORDER_BOOKS}|${CST.DB_UPDATE}|${pair}`,
					JSON.stringify(orderBookSnapshotUpdate)
				);
			return true;
		} catch (err) {
			util.logError(err);
			return false;
		}
	}

	public async getOrderBookSnapshot(pair: string) {
		const snapshotString = await redisUtil.get(
			`${CST.DB_ORDER_BOOKS}|${CST.DB_SNAPSHOT}|${pair}`
		);
		if (!snapshotString) return null;
		else return JSON.parse(snapshotString) as IOrderBookSnapshot;
	}
}
const orderBookPersistenceUtil = new OrderBookPersistenceUtil();
export default orderBookPersistenceUtil;
