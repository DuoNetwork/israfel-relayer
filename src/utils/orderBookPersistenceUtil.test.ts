import orderBookSnapshot from '../samples/test/orderBookSnapshot.json';
import orderBookPersistenceUtil from './orderBookPersistenceUtil';
import redisUtil from './redisUtil';

test('subscribeOrderBookUpdate', () => {
	redisUtil.onOrderUpdate = jest.fn();
	redisUtil.subscribe = jest.fn();
	orderBookPersistenceUtil.subscribeOrderBookUpdate('pair', (() => ({})) as any);
	expect((redisUtil.subscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('unsubscribeOrderBookUpdate', () => {
	redisUtil.unsubscribe = jest.fn();
	orderBookPersistenceUtil.unsubscribeOrderBookUpdate('pair');
	expect((redisUtil.unsubscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('publishOrderBookUpdate, with update', async () => {
	const orderBookSnapshotUpdate = {
		pair: 'pair',
		updates: [
			{
				price: 1,
				change: 2,
				count: 3,
				side: 'ask'
			}
		],
		prevVersion: 1234567890000,
		version: 1234567990000
	};
	redisUtil.publish = jest.fn(() => Promise.resolve(1));
	redisUtil.set = jest.fn(() => Promise.resolve(''));
	const res = await orderBookPersistenceUtil.publishOrderBookUpdate(
		'pair',
		orderBookSnapshot,
		orderBookSnapshotUpdate
	);
	expect((redisUtil.set as jest.Mock).mock.calls).toMatchSnapshot();
	expect((redisUtil.publish as jest.Mock).mock.calls).toMatchSnapshot();
	expect(res).toBeTruthy();
});

test('publishOrderBookUpdate, withoutsnpashot update', async () => {
	redisUtil.publish = jest.fn(() => Promise.resolve(1));
	redisUtil.set = jest.fn(() => Promise.resolve(''));
	const res = await orderBookPersistenceUtil.publishOrderBookUpdate('pair', orderBookSnapshot);
	expect((redisUtil.set as jest.Mock).mock.calls).toMatchSnapshot();
	expect(redisUtil.publish as jest.Mock).not.toBeCalled();
	expect(res).toBeTruthy();
});

test('publishOrderBookUpdate, throw Error', async () => {
	redisUtil.publish = jest.fn(() => Promise.resolve(1));
	redisUtil.set = jest.fn(() => Promise.reject());
	const res = await orderBookPersistenceUtil.publishOrderBookUpdate('pair', orderBookSnapshot);
	expect((redisUtil.set as jest.Mock).mock.calls).toMatchSnapshot();
	expect(redisUtil.publish as jest.Mock).not.toBeCalled();
	expect(res).toBeFalsy();
});

test('getOrderBookSnapshot , with result', async () => {
	redisUtil.get = jest.fn(() => Promise.resolve(JSON.stringify(orderBookSnapshot)));
	const res = await orderBookPersistenceUtil.getOrderBookSnapshot('pair');
	expect((redisUtil.get as jest.Mock).mock.calls).toMatchSnapshot();
	expect(res).toMatchSnapshot();
});

test('getOrderBookSnapshot , without result', async () => {
	redisUtil.get = jest.fn(() => Promise.resolve(''));
	const res = await orderBookPersistenceUtil.getOrderBookSnapshot('pair');
	expect((redisUtil.get as jest.Mock).mock.calls).toMatchSnapshot();
	expect(res).toMatchSnapshot();
});
