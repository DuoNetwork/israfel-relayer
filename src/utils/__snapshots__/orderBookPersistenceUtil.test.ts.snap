// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOrderBookSnapshot , with result 1`] = `
Array [
  Array [
    "orderBooks|snapshot|pair",
  ],
]
`;

exports[`getOrderBookSnapshot , with result 2`] = `
Object {
  "asks": Array [
    Object {
      "amount": 20,
      "count": 4,
      "price": 130,
    },
    Object {
      "amount": 10,
      "count": 6,
      "price": 140,
    },
    Object {
      "amount": 30,
      "count": 5,
      "price": 150,
    },
  ],
  "bids": Array [
    Object {
      "amount": 20,
      "count": 4,
      "price": 120,
    },
    Object {
      "amount": 10,
      "count": 6,
      "price": 110,
    },
    Object {
      "amount": 30,
      "count": 5,
      "price": 100,
    },
  ],
  "pair": "pair",
  "version": 1234567890000,
}
`;

exports[`getOrderBookSnapshot , without result 1`] = `
Array [
  Array [
    "orderBooks|snapshot|pair",
  ],
]
`;

exports[`getOrderBookSnapshot , without result 2`] = `null`;

exports[`publishOrderBookUpdate, throw Error 1`] = `
Array [
  Array [
    "orderBooks|snapshot|pair",
    "{\\"pair\\":\\"pair\\",\\"version\\":1234567890000,\\"bids\\":[{\\"price\\":120,\\"amount\\":20,\\"count\\":4},{\\"price\\":110,\\"amount\\":10,\\"count\\":6},{\\"price\\":100,\\"amount\\":30,\\"count\\":5}],\\"asks\\":[{\\"price\\":130,\\"amount\\":20,\\"count\\":4},{\\"price\\":140,\\"amount\\":10,\\"count\\":6},{\\"price\\":150,\\"amount\\":30,\\"count\\":5}]}",
  ],
]
`;

exports[`publishOrderBookUpdate, with update 1`] = `
Array [
  Array [
    "orderBooks|snapshot|pair",
    "{\\"pair\\":\\"pair\\",\\"version\\":1234567890000,\\"bids\\":[{\\"price\\":120,\\"amount\\":20,\\"count\\":4},{\\"price\\":110,\\"amount\\":10,\\"count\\":6},{\\"price\\":100,\\"amount\\":30,\\"count\\":5}],\\"asks\\":[{\\"price\\":130,\\"amount\\":20,\\"count\\":4},{\\"price\\":140,\\"amount\\":10,\\"count\\":6},{\\"price\\":150,\\"amount\\":30,\\"count\\":5}]}",
  ],
]
`;

exports[`publishOrderBookUpdate, with update 2`] = `
Array [
  Array [
    "orderBooks|update|pair",
    "{\\"pair\\":\\"pair\\",\\"price\\":1,\\"amount\\":2,\\"count\\":3,\\"side\\":\\"ask\\",\\"prevVersion\\":1234567890000,\\"version\\":1234567990000}",
  ],
]
`;

exports[`publishOrderBookUpdate, withoutsnpashot update 1`] = `
Array [
  Array [
    "orderBooks|snapshot|pair",
    "{\\"pair\\":\\"pair\\",\\"version\\":1234567890000,\\"bids\\":[{\\"price\\":120,\\"amount\\":20,\\"count\\":4},{\\"price\\":110,\\"amount\\":10,\\"count\\":6},{\\"price\\":100,\\"amount\\":30,\\"count\\":5}],\\"asks\\":[{\\"price\\":130,\\"amount\\":20,\\"count\\":4},{\\"price\\":140,\\"amount\\":10,\\"count\\":6},{\\"price\\":150,\\"amount\\":30,\\"count\\":5}]}",
  ],
]
`;

exports[`subscribeOrderBookUpdate 1`] = `
Array [
  Array [
    "orderBooks|update|pair",
  ],
]
`;

exports[`unsubscribeOrderBookUpdate 1`] = `
Array [
  Array [
    "orderBooks|update|pair",
  ],
]
`;
